File: NFT.py
import sys
import numpy as np

'''
Non-equidistan Fourier Transform
Uso: python3 Fourier.py "datos.dat" f0 ff step "output.dat"
el espacio de frecuencias será [f0,ff] con obvios pasos de step.
EL resultado se guardará en "output.dat"
'''

Vround=np.vectorize(round)

def DiscreteTransform(freq,x,y):
    N=len(y)
    s=[(y[i])*np.exp(-2*np.pi*1j*freq*x[i]) for i in range(N)]
    return np.sum(s)


filename=sys.argv[1]
f0=float(sys.argv[2])
ff=float(sys.argv[3])
step=float(sys.argv[4])
outname=sys.argv[5]

data=np.genfromtxt(filename)
X=data[:,0]
Y=data[:,1]
Y=Y-np.mean(Y)
fspace=np.arange(f0,ff+step,step)
sspace=[abs(DiscreteTransform(f,X,Y)) for f in fspace]

np.savetxt(outname,np.transpose(np.array([fspace,sspace])),delimiter="\t")

File: NFT.jl
#=
Non-equidistant Fourier Transform
Uso: julia NFT.jl "input.dat" f0 ff step "output.dat"
=#

I=im
len=length
str=string

#=
function SingleFT(t::Float64, y::Float64, f::Float64)
    """
    Single iteration of fourier transform
    """
    return y*exp(-2.0*pi*I*f*t)
end
=#

SFT=(t,y,f)->y*exp(-2.0*pi*I*f*t)

filename=ARGS[1]
f0=float(ARGS[2])
ff=float(ARGS[3])
step=float(ARGS[4])
outname=ARGS[5]

file=readlines(open(filename))
X=zeros(length(file))
Y=zeros(length(file))
for i in 1:length(file)
    X[i]+=float( split(file[i])[1] )
    Y[i]+=float( split(file[i])[2] )
end
Y=Y-mean(Y)

#=
transf=[]

for freq in f0:step:ff
    FT=(t,y)->SFT(t,y,freq)
    push!(transf,sum(map(FT,X,Y)))
end
=#


open(outname, "w") do outf
    for freq in f0:step:ff
        FT=(t,y)->SFT(t,y,freq)
        s = abs(sum(map(FT,X,Y)))
        write(outf, str(freq)*"\t"*str(s)*"\n" )
    end
end
